File: backend/UsersPool/function.js
const AWS = require('aws-sdk');
const ssm = new AWS.SSM();
const cognito = new AWS.CognitoIdentityServiceProvider();

exports.handler = async (event, context) => {
    console.log('Event:');
    console.log(JSON.stringify(event));
    
    const { RequestType, ResourceProperties } = event;
    const userPoolId = ResourceProperties.UserPoolId;
    const clientId = ResourceProperties.UserPoolClientId;
    const parameterName = ResourceProperties.ParameterName;

    try {
        if (RequestType === 'Create' || RequestType === 'Update') {
            // Get the User Pool Client details
            const userPoolClient = await cognito.describeUserPoolClient({
                UserPoolId: userPoolId,
                ClientId: clientId
            }).promise();

            const clientSecret = userPoolClient.UserPoolClient.ClientSecret; 

            // Write the Client Secret to SSM Parameter Store

            await ssm.putParameter({
                Name: parameterName +"/userPoolId",
                Value: userPoolId,
                Type: 'SecureString',
                Overwrite: true
            }).promise();

            await ssm.putParameter({
                Name: parameterName +"/userPoolClientId",
                Value: clientId,
                Type: 'SecureString',
                Overwrite: true
            }).promise();

            await ssm.putParameter({
                Name: parameterName +"/userPoolClientSecret",
                Value: clientSecret,
                Type: 'SecureString',
                Overwrite: true
            }).promise();

           

            // Send a response back to CloudFormation indicating success
            return sendResponse(event, context, 'SUCCESS',clientSecret);
        } else if (RequestType === 'Delete') {
            try{
                // If it's a Delete event, you may choose to delete the SSM parameter if needed
                await ssm.deleteParameter({
                   Name: parameterName +"/userPoolId"
                }).promise();

                await ssm.deleteParameter({
                   Name: parameterName +"/userPoolClientId"
                }).promise();

                await ssm.deleteParameter({
                    Name: parameterName +"/userPoolClientSecret"
                 }).promise();
            }catch(ignore){
                console.error('Failed to delete parameter:', ignore.message);
            }

            // Send a response back to CloudFormation indicating success
            return sendResponse(event, context, 'SUCCESS',"");
        } else {
            console.error('Error: Invalid RequestType');
            return sendResponse(event, context, 'SUCCESS',"");
        }
    } catch (error) {
        console.error('Error:', error.message);        
        //do not fail the CFN stack
        return sendResponse(event, context, 'SUCCESS',"");
    }
};

function sendResponse(event, context, responseStatus,clientSecret) {
    var response = require('cfn-response');
    response.send(event, context, responseStatus, {
        ClientSecret: clientSecret
    }); 
}

File: backend/UsersPool/function.py
import boto3
import cfnresponse

ssm = boto3.client('ssm')
cognito = boto3.client('cognito-idp')

def handler(event, context):
    print('Event:')
    print(event)
    
    request_type = event['RequestType']
    resource_properties = event['ResourceProperties']
    user_pool_id = resource_properties['UserPoolId']
    client_id = resource_properties['UserPoolClientId']
    parameter_name = resource_properties['ParameterName']
    try:
        if request_type == 'Create' or request_type == 'Update':
            # Get the User Pool Client details
            user_pool_client = cognito.describe_user_pool_client(
                UserPoolId=user_pool_id,
                ClientId=client_id
            )

            client_secret = user_pool_client['UserPoolClient']['ClientSecret']

            # Write the Client Secret to SSM Parameter Store
            put_parameter(parameter_name + "/userPoolId",user_pool_id)
            put_parameter(parameter_name + "/userPoolClientId",client_id)    
            put_parameter(parameter_name + "/userPoolClientSecret",client_secret)

            # Send a response back to CloudFormation indicating success
            return send_response(event, context, 'SUCCESS', client_secret)
        elif request_type == 'Delete':
        
                # If it's a Delete event, you may choose to delete the SSM parameter if needed
                delete_parameter(parameter_name + "/userPoolId")
                delete_parameter(parameter_name + "/userPoolClientId")
                delete_parameter(parameter_name + "/userPoolClientSecret")
                
                return send_response(event, context, 'SUCCESS')
    except Exception as e:
        print(e)
        return send_response(event, context, 'FAILED')

def delete_parameter(name):
    try:
        ssm.delete_parameter(Name=name)
    except Exception as e:
        print(e)      

def put_parameter(name, value):
    try:
        ssm.put_parameter(
            Name=name,
            Value=value,
            Type="SecureString",
            Overwrite=True
        )
    except Exception as e:
        print(e)
     

def send_response(event, context, response_status, client_secret=""):             
    responseData = {}
    responseData['Data'] = client_secret
    cfnresponse.send(event, context, cfnresponse.SUCCESS, responseData, "CustomResourcePhysicalID")